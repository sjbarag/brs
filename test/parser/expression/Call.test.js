const BrsError = require("../../../lib/Error");
const { Lexeme, BrsTypes, Parser } = require("brs");
const { Int32 } = BrsTypes;

const { EOF } = require("../ParserTests");

describe("parser call expressions", () => {
    afterEach(() => BrsError.reset());

    it("parses named function calls", () => {
        const parsed = Parser.parse([
            { kind: Lexeme.Identifier, text: "RebootSystem", line: 1 },
            { kind: Lexeme.LeftParen,  text: "(", line: 1 },
            { kind: Lexeme.RightParen, text: ")", line: 1 },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });

    it("allows closing parentheses on separate line", () => {
        const parsed = Parser.parse([
            { kind: Lexeme.Identifier, text: "RebootSystem", line: 1 },
            { kind: Lexeme.LeftParen,  text: "(", line: 1 },
            { kind: Lexeme.Newline, text: "\\n", line: 1 },
            { kind: Lexeme.Newline, text: "\\n", line: 2 },
            { kind: Lexeme.RightParen, text: ")", line: 3 },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed).toMatchSnapshot();
    });

    it("accepts arguments", () => {
        const parsed = Parser.parse([
            { kind: Lexeme.Identifier, text: "add", line: 1 },
            { kind: Lexeme.LeftParen,  text: "(", line: 1 },
            { kind: Lexeme.Integer, text: "1", literal: new Int32(1) },
            { kind: Lexeme.Comma,  text: ",", line: 1 },
            { kind: Lexeme.Integer, text: "2", literal: new Int32(2) },
            { kind: Lexeme.RightParen, text: ")", line: 1 },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();
        expect(parsed).not.toBeNull();
        expect(parsed[0].expression.args).toBeTruthy();
        expect(parsed).toMatchSnapshot();
    });
});
