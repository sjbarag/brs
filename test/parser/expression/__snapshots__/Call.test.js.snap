// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser call expressions accepts arguments 1`] = `
Array [
  Expression {
    "expression": Call {
      "args": Array [
        Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Int32 {
            "kind": 4,
            "value": 1,
          },
        },
        Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Int32 {
            "kind": 4,
            "value": 2,
          },
        },
      ],
      "callee": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "add",
        },
        "type": "Variable",
      },
      "closingParen": Object {
        "isReserved": false,
        "kind": "RightParen",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": ")",
      },
      "type": "Call",
    },
    "type": "Expression",
  },
]
`;

exports[`parser call expressions allows closing parentheses on separate line 1`] = `
Array [
  Expression {
    "expression": Call {
      "args": Array [],
      "callee": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "RebootSystem",
        },
        "type": "Variable",
      },
      "closingParen": Object {
        "isReserved": false,
        "kind": "RightParen",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": ")",
      },
      "type": "Call",
    },
    "type": "Expression",
  },
]
`;

exports[`parser call expressions does not invalidate the next statement on a multi-statement line 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 5,
          },
          "file": "",
          "start": Object {
            "column": 28,
            "line": 2,
          },
        },
        "statements": Array [
          Assignment {
            "name": Object {
              "isReserved": false,
              "kind": "Identifier",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": 28,
                  "line": 4,
                },
                "file": "",
                "start": Object {
                  "column": 24,
                  "line": 4,
                },
              },
              "text": "name",
            },
            "tokens": Object {
              "equals": Object {
                "isReserved": false,
                "kind": "Equal",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 30,
                    "line": 4,
                  },
                  "file": "",
                  "start": Object {
                    "column": 29,
                    "line": 4,
                  },
                },
                "text": "=",
              },
            },
            "type": "Assignment",
            "value": Literal {
              "_location": Object {
                "end": Object {
                  "column": 36,
                  "line": 4,
                },
                "file": "",
                "start": Object {
                  "column": 31,
                  "line": 4,
                },
              },
              "type": "Literal",
              "value": BrsString {
                "kind": 3,
                "value": "bob",
              },
            },
          },
        ],
        "type": "Block",
      },
      "endKeyword": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 5,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 5,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Expr_Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 26,
          "line": 2,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 2,
        },
      },
      "text": "DoThingOne",
    },
    "type": "Stmt_Function",
  },
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 7,
          },
          "file": "",
          "start": Object {
            "column": 28,
            "line": 6,
          },
        },
        "statements": Array [],
        "type": "Block",
      },
      "endKeyword": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 7,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 7,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 6,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 6,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Expr_Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 26,
          "line": 6,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 6,
        },
      },
      "text": "DoThingTwo",
    },
    "type": "Stmt_Function",
  },
]
`;

exports[`parser call expressions does not invalidate the rest of the file on incomplete statement 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 4,
          },
          "file": "",
          "start": Object {
            "column": 28,
            "line": 2,
          },
        },
        "statements": Array [],
        "type": "Block",
      },
      "endKeyword": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 4,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 4,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Expr_Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 26,
          "line": 2,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 2,
        },
      },
      "text": "DoThingOne",
    },
    "type": "Stmt_Function",
  },
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 6,
          },
          "file": "",
          "start": Object {
            "column": 28,
            "line": 5,
          },
        },
        "statements": Array [],
        "type": "Block",
      },
      "endKeyword": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 6,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 6,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 5,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 5,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Expr_Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 26,
          "line": 5,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 5,
        },
      },
      "text": "DoThingTwo",
    },
    "type": "Stmt_Function",
  },
]
`;

exports[`parser call expressions parses named function calls 1`] = `
Array [
  Expression {
    "expression": Call {
      "args": Array [],
      "callee": Variable {
        "name": Object {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "RebootSystem",
        },
        "type": "Variable",
      },
      "closingParen": Object {
        "isReserved": false,
        "kind": "RightParen",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": ")",
      },
      "type": "Call",
    },
    "type": "Expression",
  },
]
`;
